src/Tools/dotnet-monitor/Auth/ApiKeyAuthenticationHandler.cs

```diff
@@ -0,0 +1,99 @@
+
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Diagnostics.Monitoring.RestServer;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+using Microsoft.Net.Http.Headers;
+using System;
+using System.Buffers.Text;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using System.Security.Cryptography;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor
+{
+    /// <summary>
+    /// Authenticates against the ApiKey stored on the server.
+    /// </summary>
+    internal sealed class ApiKeyAuthenticationHandler : AuthenticationHandler<ApiKeyAuthenticationHandlerOptions>
+    {
+        private const string AuthConfigurationKey = "ApiAuthentication:ApiKeyHash";
+
+        public ApiKeyAuthenticationHandler(IOptionsMonitor<ApiKeyAuthenticationHandlerOptions> options, ILoggerFactory loggerFactory, UrlEncoder encoder, ISystemClock clock)
+            : base(options, loggerFactory, encoder, clock)
+        {
+        }
+
+        protected override Task<AuthenticateResult> HandleAuthenticateAsync()
+        {
+            //We are expecting a header such as Authorization: <Schema> <key>
+            //If this is not present, we will return NoResult and move on to the next authentication handler.
+            if (!Request.Headers.TryGetValue(HeaderNames.Authorization, out StringValues values) ||
+                !values.Any())
+            {
+                return Task.FromResult(AuthenticateResult.NoResult());
+            }
+
+            if (!AuthenticationHeaderValue.TryParse(values.First(), out AuthenticationHeaderValue authHeader))
+            {
+                return Task.FromResult(AuthenticateResult.Fail("Invalid authentication header"));
+            }
+
+            if (!string.Equals(authHeader.Scheme, Scheme.Name, StringComparison.OrdinalIgnoreCase))
+            {
+                return Task.FromResult(AuthenticateResult.NoResult());
+            }
+
+            //The user is passing a base 64-encoded version of the secret
+            //We will be hash this and compare it to the secret in our configuration.
+            byte[] secret = new byte[32];
+            Span<byte> span = new Span<byte>(secret);
+            if (!Convert.TryFromBase64String(authHeader.Parameter, span, out int bytesWritten) || bytesWritten < 32)
+            {
+                return Task.FromResult(AuthenticateResult.Fail("Invalid Api Key format"));
+            }
+
+            using HashAlgorithm algorithm = SHA256.Create();
+            byte[] hashedSecret = algorithm.ComputeHash(secret);
+
+            IConfiguration config = Context.RequestServices.GetRequiredService<IConfiguration>();
```

Comment:
We may need an updated Configuration package to make sure that we always get the latest secret, even if the user updates the secret in scenarios such as Kubernetes.